### ISSUE TITLE
Major lag in switching focus between windows when input method is not shared between them

### ISSUE THREAD

**Issue by maxigaz** [ibus/ibus#2141]
Please fill in the following items if you don't know the root cause.

Which distribution and version?:
Manjaro Linux (64-bit)

Which desktop environment and version?:
i3 4.17.1

Which session type?:
X11

IBus version?:
1.5.21

Issue description:
I experience major lags when in ibus the option “Share the same input method among all applications” is unchecked and I switch window focus rapidly multiple times in a row. By lags, I mean that shifting from one window to another is delayed and the system temporarily becomes unresponsive.

(It’s also worth mentioning that I have a rather slow CPU, a Dual Core Celeron G1610. With a newer, more powerful CPU, the amount of lag may be different.)

Steps to reproduce:
1. Run `ibus-dameon -drx`.
1. Open several windows of various applications (I use Firefox and Kitty for testing).
1. Right click on the panel icon of ibus, click on Preferences, and under Advanced, and make sure “Share the same input method among all applications” is unchecked.
1. Switch focus between windows more than one time in a row quickly.
1. All except the first focus switch has a noticable delay. During this delay, the application window you’re trying to get focus to also becomes unresponsive.

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
(Run `ibus exit` and `ibus-daemon --xim &`)
Yes.

Do you see any errors when you run ibus-daemon with the verbose option?:
(Run `ibus-daemon --xim --verbose &` and look at the output when you encounter your problem.)
I can see some messages when I run the above command, but not when I experience the problem itself.

```
(ibus-ui-gtk3:859): IBUS-WARNING **: 12:03:01.416: panel.vala:255: If you launch KDE5 on xterm, export XDG_CURRENT_DESKTOP=KDE before launch KDE5.

(ibus-ui-gtk3:859): IBUS-WARNING **: 12:03:01.506: ibus_bus_call_sync: org.freedesktop.DBus.Properties.Get: GDBus.Error:org.freedesktop.DBus.Error.Failed: No global engine.

(ibus-ui-gtk3:859): Gdk-CRITICAL **: 12:03:01.914: gdk_window_thaw_toplevel_updates: assertion 'window->update_and_descendants_freeze_count > 0' failed
```

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
I haven’t tried.

**Comment by fujiwarat** [comment]
I cannot find Kitty in Fedora.
Can you reproduce your problem with Firefox and gnome-terminal?
Which versions of firefox and gnome-terminal do you use?
Since I don't use this mode, the priority would be low.

**Comment by maxigaz** [comment]
Yes, I can reproduce it with Firefox 69.0.3 and GNOME Terminal 3.34.2. (I click inside the URL bar in Firefox, hit `Ctrl-Space` to switch from the default layout, Hungarian, to Japanese (through `ibus-kkc`), and then focus switching between the two windows slows down.)

By the way, according to the [homepage of Kitty](https://sw.kovidgoyal.net/kitty/), it is actually available for Fedora; it points to [this page](https://copr.fedorainfracloud.org/coprs/gagbo/kitty-latest/).

**Comment by vanphong1310** [comment]
Same problem here. I use Linux Mint 20 Cinnamon with iBus 1.5.22 with two input method, when I switch off 'Share same input method among all applications' option of IBus, not only Firefox but all windows include popups, widgets, ... slows down immediately, minimize or maximize a window take 5s.

**Comment by vanphong1310** [comment]
Hello @fujiwarat,
I have tested my system with fcitx (two input methods like IBus), there are no problems with Share state among Window feature.
I used IBus for a long time with Linux Mint 19, everything works fine. After fresh install Linux Mint 20, install IBus and switch off 'Share same input method among all applications' option, my system is so laggy.
I noticed that Linux Mint 20 have update X.Org X Server version to support NVIDIA "On-Demand" feature out of the box. When I turn on 'Share same input method among all applications' option and minimize/maximize a text edit window multiple times, the Xorg process take only 8% cpu usage (top command output on 8 core), but if I turn off this option, the Xorg process take up to 70% cpu usage. So I assume that IBus has a problem with new Xorg version.
My X.Org X Server version:
`$ xdpyinfo | grep version`
`version number:    11.0`
`X.Org version: 1.20.8`
Can you check it? Please let me know if you need any further information.

**Comment by WGH-** [comment]
The corresponding bug report on the i3 side: https://github.com/i3/i3/issues/3924

**Comment by fujiwarat** [comment]
I still cannot reproduce this issue.
How long did you play to reproduce it?

**Comment by vanphong1310** [comment]
@fujiwarat Thanks for your attention!
The problem appears when I turn off "Use system keyboard layout" option in Advanced tab of IBus Preferences. I turned it on and lag is not any more now. "Share same input method among all applications" option is not cause but it make the problem worse.
I use IBus 1.5.22 on Linux Mint 20.1
